flower_with_bonsai.C

```diff
@@ -0,0 +1,185 @@
+		return -1;
+	}
+
+	// Read geometry from ROOT file and initialize Bonsai
+	TTree *geotree = (TTree*)file->Get("wcsimGeoT");
+	WCSimRootGeom *geo = 0;
+	geotree->SetBranchAddress("wcsimrootgeom", &geo);
+	if (geotree->GetEntries() == 0) exit(9); // exit if no geometry is defined in the ROOT file
+	geotree->GetEntry(0);
+	bonsai->Init(geo);
+
+	WCSimFLOWER * flower = new WCSimFLOWER(detector, geo, overwrite_nearest, verbose);
+
+	TTree *tree = (TTree*)file->Get("wcsimT"); // Get a pointer to the tree from the file
+	WCSimRootEvent* event = new WCSimRootEvent(); // Create WCSimRootEvent to put stuff from the tree in
+	tree->SetBranchAddress("wcsimrootevent", &event); // Set branch address for reading from tree
+	tree->GetBranch("wcsimrootevent")->SetAutoDelete(kTRUE); // Force deletion to prevent memory leak
+	WCSimRootTrigger* trigger; // will contain triggers of event later (0: initial particle; 1..n: decay products)
+
+	int ncherenkovdigihits, i;
+	double x, y, z, eRec;	
+
+	// used by bonsai->BonsaiFit()
+	float bsVertex[4],bsResult[6], bsGood[3];
+	int *bsNhit;
+	int bsNsel[2];
+
+	// Loop over events
+	cout << "Starting timer for the event loop" << endl;
+	TStopwatch timer;
+	for (int ev=0; ev < tree->GetEntries(); ev++) {
+		if (verbose) std::cout << "event number: " << ev << std::endl;
+
+		// Read the event from the tree into the WCSimRootEvent instance
+		tree->GetEntry(ev);
+		trigger = event->GetTrigger(0);
+
+		// See chapter 5 of doc/DetectorDocumentation.pdf in the WCSim repository
+		// for more information on the structure of the root file.
+
+		// Loop over triggers in the event
+		for (int index = 0 ; index < event->GetNumberOfEvents(); index++) {
+			// We don't have secondary decays so shouldn't observe more than 1 trigger per event.
+			// If we do, it's noise (because the NDigits trigger in WCSim is not optimized for HK).
+			// Those would get filtered later in energy cuts, but for simplicity, we ignore them here.
+			if (index == 1) continue;
+
+			trigger = event->GetTrigger(index);
+			ncherenkovdigihits = trigger->GetNcherenkovdigihits();
+			if (verbose) std::cout << "ncherenkovdigihits: " << ncherenkovdigihits << std::endl;
+			if (ncherenkovdigihits == 0) {
+				std::cout << "t, PID, 0, 0, 0, 0" << std::endl;
+				continue;
+			}
+			std::vector<float> bsT (ncherenkovdigihits,0);
+			std::vector<float> bsQ (ncherenkovdigihits,0);
+			std::vector<int> bsCAB (ncherenkovdigihits,0);
+
+			bsNhit = & ncherenkovdigihits;
+
+			// Get time, charge and PMT number for each hit
+			for (i=0; i<ncherenkovdigihits; i++) {
+				TObject *element = (trigger->GetCherenkovDigiHits())->At(i);
+				WCSimRootCherenkovDigiHit *cherenkovdigihit = dynamic_cast<WCSimRootCherenkovDigiHit*>(element);
+
+				bsT[i] = cherenkovdigihit->GetT();
+				bsQ[i] = cherenkovdigihit->GetQ();
+				bsCAB[i] = cherenkovdigihit->GetTubeId();
+
+				if(bsCAB[i] == 0)
+				  cout << "WTF!?!? Digit has tube ID 0" << endl;
```

Comment:
This looks a bit like some debug code that slipped through? Would be good to either delete this or have a somewhat more helpful output.

Reply:
Made more helpful